src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1959,6 +2009,139 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 
 
 /*
+ * HasNonPartitionColumnDistinctAgg returns true if target entry or having qualifier
+ * has non-partition column reference in aggregate definition. Note that, it only
+ * checks aggs subfield of Aggref, it does not check FILTER or SORT clauses.
+ */
+static bool
+HasNonPartitionColumnDistinctAgg(List *targetEntryList, Node *havingQual,
+								 List *tableNodeList)
+{
+	List *targetVarList = pull_var_clause((Node *) targetEntryList,
+										  PVC_INCLUDE_AGGREGATES);
+	List *havingVarList = pull_var_clause((Node *) havingQual, PVC_INCLUDE_AGGREGATES);
+	List *aggregateCheckList = list_concat(targetVarList, havingVarList);
+	ListCell *aggregateCheckCell = NULL;
+	foreach(aggregateCheckCell, aggregateCheckList)
+	{
+		Node *targetNode = lfirst(aggregateCheckCell);
+		Aggref *targetAgg = NULL;
+		List *varList = NIL;
+		ListCell *varCell = NULL;
+		bool isPartitionColumn = false;
+
+		if (IsA(targetNode, Var))
+		{
+			continue;
+		}
+
+		Assert(IsA(targetNode, Aggref));
+		targetAgg = (Aggref *) targetNode;
+		if (targetAgg->aggdistinct == NIL)
+		{
+			continue;
+		}
+
+		varList = pull_var_clause_default((Node *) targetAgg->args);
+
+		if (varList == NIL)
+		{
+			continue;
+		}
+
+		foreach(varCell, varList)
+		{
+			Node *targetVar = (Node *) lfirst(varCell);
+			Assert(IsA(targetVar, Var));
+
+			isPartitionColumn = ExpressionContainsPartitionColumn(
+				targetVar, tableNodeList);
+
+			if (!isPartitionColumn)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ShouldPullDistinctColumn returns true if distinct aggregate should pull
+ * individual columns from worker to master and evaluate aggregate operation
+ * at master.
+ */
+static bool
+ShouldPullDistinctColumn(bool groupedByDisjointPartitionColumn,
```

Comment:
This function is a bit confusing. Could you describe in which conditions we want to pull columns to master and run aggregate just on the master node? Maybe also refactor a bit?

Reply:
And is this `PullDistinctColums` or `PullColumns` because if `repartitionSubquery` we just return true.

Reply:
it pulls only distinct columns, if it is repartition query it should pull columns regardless of partition type or group by.

Reply:
added pull cases